### PR TITLE
[Fixes #9007] correct wiring between ember-cli's watcher and bro…

### LINKED ISSUE #9007: ember-cli >= 3.15 or EMBER_CLI_BROCCOLI_WATCHER=true prior to 3.15 usage does not correctly print out errors
In 3.14, an app.import of a missing file was an immediate exit with an expected stack trace. In 3.15, it is now a hung build with a UnhandledPromiseRejectionWarning.

3.14 https://travis-ci.org/kellyselden/ember-cli-test-server/jobs/633006370#L624
3.15 https://travis-ci.org/kellyselden/ember-cli-test-server/jobs/633037017#L560

My test suite is intentionally causing this error, that's the only way I discovered this. Here is the line
https://github.com/kellyselden/ember-cli-test-server/blob/764e703573/test/index-test.js#L112
https://github.com/kellyselden/ember-cli-test-server/blob/764e703573/test/index-test.js#L61

### PR DESCRIPTION
**Pull request by stefanpenner**
…ccoli's watcher

TL;DR broccoli's watcher emits both `error` and `buildFailure` events as it differentiates between the two.  We accidentally only subscribed to `error`.

This  PR corrects this issue.

### CODE REVIEW

File: lib/models/watcher.js [hunk 1]
```diff
@@ -30,7 +30,11 @@ module.exports = class Watcher extends CoreObject {
     this.watcher.on('buildStart', this._setupBroccoliWatcherBuild.bind(this));
     this.watcher.on('buildSuccess', this.didChange.bind(this));
 
+    // build errors arriving via watcher
+    this.watcher.on('buildFailure', this.didError.bind(this));
```
Comment (stefanpenner) [inline lib/models/watcher.js]: the actual fix

File: lib/models/watcher.js [hunk 2]
```diff
@@ -55,22 +59,19 @@ module.exports = class Watcher extends CoreObject {
...
-          // remove the heimdall subtree for this build so we don't leak.  If
-          // BROCCOLI_VIZ=1 then we have already output the json in `verboseOutput`.
-          heimdallNode.remove();
-        }
-      });
+  async _setupBroccoliWatcherBuild() {
+    try {
+      const hash = await this.watcher.currentBuild;
+      hash.graph.__heimdall__.remove();
+    } catch (e) {
```
Comment (stefanpenner) [inline lib/models/watcher.js]: ensure we correctly handle errors off this instance of `currentBuild`

File: tests/acceptance/smoke-test-slow.js [hunk 3]
```diff
@@ -328,6 +328,54 @@ describe('Acceptance: smoke-test', function() {
     }
   });
 
+  it('build failures should be logged correctly', async function() {
```
Comment (stefanpenner) [inline tests/acceptance/smoke-test-slow.js]: acceptance test for the motivating problem

File: tests/helpers/run-command.js [hunk 4]
```diff
@@ -43,18 +44,21 @@ module.exports = function run(/* command, args, options */) {
     options.log(`      Running: ${command} ${args.join(' ')} in: ${process.cwd()}`);
 
     let opts = {};
+    args = [`--unhandled-rejections=strict`, `${command}`].concat(args);
```
Comment (stefanpenner) [inline tests/helpers/run-command.js]: ensure all acceptance tests that spawn node sub-processes via `run-command` also run in the same `--unhandled-rejections=strict` mode

File: lib/models/watcher.js [hunk 5]
```diff
@@ -55,22 +59,19 @@ module.exports = class Watcher extends CoreObject {
...
+      if (e !== null && typeof e === 'object' && e.isBuilderError === true) {
+        // it is safe to ignore build errors here, as this code only exists to
+        // prune heimdall data after a build.
```
Comment (mike-north) [inline lib/models/watcher.js]: is it intentional to leave an empty block here? How do you feel about the "minimizing WTFs" factor compared to
```js
const isExpctedError = e !== null && typeof e === 'object' && e.isBuilderError === true; 
if (!isExpctedError) throw e;
// explanation of why we're safe here
```

File: tests/helpers/run-command.js [hunk 6]
```diff
@@ -43,18 +44,21 @@ module.exports = function run(/* command, args, options */) {
...
-      command = 'node';
       opts.windowsVerbatimArguments = true;
       opts.stdio = [null, null, null, 'ipc'];
     }
     if (options.env) {
       opts.env = defaults(options.env, process.env);
     }
 
-    debug.info('command: %s, args: %o', command, args);
+    console.log('runCommand: %s, args: %o', command, args);
```
Comment (mike-north) [inline tests/helpers/run-command.js]: I believe this change will result in things being logged in test output. Is this desirable?

File: lib/models/watcher.js [hunk 7]
```diff
@@ -55,22 +59,19 @@ module.exports = class Watcher extends CoreObject {
...
+      } else {
+        // this error is unexpected
+        throw e;
```
Comment (mike-north) [inline lib/models/watcher.js]: IIRC re-throwing like this will not add `_setupBroccoliWatcherBuild` to the stack trace. Would it be useful for debugging purposes to know that this frame was involved in the failure?

File: tests/helpers/run-command.js (hunk 6 above)
Comment (rwjblue) [reply tests/helpers/run-command.js]: No, I don’t think we should change this permanently

File: tests/helpers/run-command.js [hunk 8]
```diff
@@ -43,18 +44,21 @@ module.exports = function run(/* command, args, options */) {
...
     let child = spawn(command, args, opts);
+    // ensure we tear down the child process on exit;
+    captureExit.onExit(() => killCliProcess(child));
```
Comment (rwjblue) [inline tests/helpers/run-command.js]: I’d prefer you to change the line above to use `execa` which automatically takes care of this bit for us

File: lib/models/watcher.js (hunk 5 above)
Comment (stefanpenner) [reply lib/models/watcher.js]: Yes, this is intentional. My preference is the clear `if` `else` in this cases, although I can improve the comment further.

File: lib/models/watcher.js (hunk 7 above)
Comment (stefanpenner) [reply lib/models/watcher.js]: > Would it be useful for debugging purposes to know that this frame was involved in the failure?

Not particularly, especially considering the complexity of stitching the stack frames together.

File: tests/helpers/run-command.js (hunk 8 above)
Comment (stefanpenner) [reply tests/helpers/run-command.js]: I would prefer to not couple switching to `execa` as part of this fix. I don't mind experimenting with that in a separate/later PR.

Comment (rwjblue) [reply tests/helpers/run-command.js]: Ok, seems good to me. Mind making an issue at least (so between the two of us we remember to circle back).

**Comment by rwjblue** [comment]
Looks like this needs a rebase

File: tests/helpers/run-command.js (hunk 8 above)
Comment (stefanpenner) [reply tests/helpers/run-command.js]: https://github.com/ember-cli/ember-cli/issues/9073

**Comment by stefanpenner** [comment]
@rwjblue this will likely need to be back-ported too 3.15

**Comment by rwjblue** [comment]
All of the extended Travis jobs failed for:

```
not ok 1369 Acceptance: smoke-test build failures should be logged correctly
2068  the object {
2069    "code": 1
2070    "errors": [
2071      "Port 54323 is already in use.\n"
2072      "\n"
2073    ]
2074    "output": []
2075    "signal": [null]
2076  } was thrown, throw an Error :)
2077  Error: the object {
2078    "code": 1
2079    "errors": [
2080      "Port 54323 is already in use.\n"
2081      "\n"
2082    ]
2083    "output": []
2084    "signal": [null]
2085  } was thrown, throw an Error :)
```

**Comment by rwjblue** [comment]
RE: backport, 3.15 is unsupported at this time, once this lands I’ll target these bugfixes to release branch and get them out in a 3.16 patch release (3.16 is the current release, and soon will be an LTS release).

**Comment by stefanpenner** [comment]
Ya, I noticed that before I went to sleep. I’ll take a look todG

File: tests/helpers/run-command.js [hunk 9]
```diff
@@ -111,4 +94,15 @@ module.exports = function run(/* command, args, options */) {
       }
     });
   });
+
+  promise.kill = function() {
```
Comment (stefanpenner) [inline tests/helpers/run-command.js]: A broader refactoring of `runCommand` is needed, but I worry about doing that in this PR as we intend to as a bugfix + backport.

Tracking `runCommand` related improvements here: https://github.com/ember-cli/ember-cli/issues/9073#issue-573121155

**Comment by stefanpenner** [comment]
I'm pretty sure I know the issue, I'll be digging in next time I have free cycles.

File: tests/acceptance/brocfile-smoke-test-slow.js [hunk 10]
```diff
@@ -33,6 +33,7 @@ describe('Acceptance: brocfile-smoke-test', function() {
   });
 
   afterEach(function() {
+    runCommand.killAll();
```
Comment (stefanpenner) [inline tests/acceptance/brocfile-smoke-test-slow.js]: interim solution to ensure we kill any leaked `runCommand` processes, once this PR lands. I will invest time in improving this area in https://github.com/ember-cli/ember-cli/issues/9073 after this PR lands.

File: tests/helpers/run-command.js [hunk 11]
```diff
@@ -6,7 +6,9 @@ const defaults = require('ember-cli-lodash-subset').defaults;
 const killCliProcess = require('./kill-cli-process');
 const logOnFailure = require('./log-on-failure');
 let debug = require('heimdalljs-logger')('run-command');
+const captureExit = require('capture-exit');
 
+let RUNS = [];
```
Comment (rwjblue) [inline tests/helpers/run-command.js]: I don't understand how this works. AFAICT nothing pushes stuff into `RUNS` (and it is not exported), am I just missing something?

File: tests/unit/models/builder-test.js [hunk 12]
```diff
@@ -15,7 +15,7 @@ let expect = chai.expect;
 let file = chai.file;
 
 let root = process.cwd();
-let tmproot = path.join(root, 'tmp');
+let tmpRoot = path.join(root, 'tmp');
```
Comment (rwjblue) [inline tests/unit/models/builder-test.js]: Seems fine, but I don't really understand the need for this change

File: tests/helpers/run-command.js (hunk 11 above)
Comment (stefanpenner) [reply tests/helpers/run-command.js]: we export a helper `killAll` which iterates this structure.

Comment (stefanpenner) [reply tests/helpers/run-command.js]: good catch!
